### PR TITLE
[Feature] qlora support

### PR DESCRIPTION
**Pull request by linsj20**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

qlora feature support

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/quantization/__init__.py
Comment (ver217) [inline colossalai/booster/quantization/__init__.py:1]: Can we move all these to `colossalai/quantization`?

Comment (linsj20) [reply colossalai/booster/quantization/__init__.py:1]: Done.

File: requirements/requirements-test.txt [hunk 1]
```diff
@@ -19,4 +19,5 @@ ninja
 flash_attn==2.0.5
 datasets
 peft>=0.7.1
+bitsandbytes>=0.39.0
```
Comment (ver217) [inline requirements/requirements-test.txt:22]: This is not needed as it's in requirements.txt

File: tests/test_lora/test_lora.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            dist.barrier()
+
+            # The Lora checkpoint should be small in size
+            checkpoint_size_mb = os.path.getsize(os.path.join(lora_ckpt_path, "adapter_model.bin")) / (1024 * 1024)
+            assert checkpoint_size_mb < 1
+
+            model_load = booster.enable_lora(model_load, pretrained_dir=lora_ckpt_path, **test_config)
+            model_load, _, _, _, _ = booster.boost(model_load)
+
+            check_state_dict_equal(model_save.state_dict(), model_load.state_dict())
```
Comment (ver217) [inline tests/test_lora/test_lora.py:105]: As these two function have many common codes. Can we merge them into one? E.g. we save & load lora checkpoint before training.

Comment (linsj20) [reply tests/test_lora/test_lora.py]: Merged.

**Review by ver217 — APPROVED** [review]

**Pull request merged**